incubator/binding-pgsql/src/main/java/io/aklivity/zilla/runtime/binding/pgsql/internal/stream/PgsqlClientFactory.java

```diff
@@ -0,0 +1,1662 @@
+    }
+
+    private int decodePgsqlMessage(
+        PgsqlClient client,
+        long traceId,
+        long authorization,
+        long budgetId,
+        DirectBuffer buffer,
+        int offset,
+        int limit)
+    {
+        int progress = offset;
+
+        final PgsqlMessageFW pgsqlMessage = messageRO.tryWrap(buffer, offset, limit);
+
+        if (pgsqlMessage != null)
+        {
+            final int type = pgsqlMessage.type();
+
+            final PgsqlClientDecoder decoder = decodersByType.getOrDefault(type, decodePgsqlIgnoreOne);
+            client.decoder = decoder;
+        }
+
+        return progress;
+    }
+
+    private int decodePgsqlMessageAuth(
+        PgsqlClient client,
+        long traceId,
+        long authorization,
+        long budgetId,
+        DirectBuffer buffer,
+        int offset,
+        int limit)
+    {
+        int progress = offset;
+
+        final PgsqlAuthenticationMessageFW pgsqlAuth = authMessageRO.tryWrap(buffer, offset, limit);
+
+        if (pgsqlAuth != null && pgsqlAuth.authenticationType() == 0)
+        {
+            client.decoder = decodePgsqlBackendKey;
+            progress = pgsqlAuth.limit();
+        }
+
+        return progress;
+    }
+
+    private int decodePgsqlMessageBackendKey(
+        PgsqlClient client,
+        long traceId,
+        long authorization,
+        long budgetId,
+        DirectBuffer buffer,
+        int offset,
+        int limit)
+    {
+        int progress = offset;
+
+        final PgsqlBackendKeyMessageFW pgsqlBackendKey = backendKeyMessageRO.tryWrap(buffer, offset, limit);
+
+        if (pgsqlBackendKey != null)
+        {
+            client.decoder = decodePgsqlParameterStatus;
+            progress = pgsqlBackendKey.limit();
+        }
+
+        return progress;
+    }
+
+    private int decodePgsqlMessageParameterStatus(
+        PgsqlClient client,
+        long traceId,
+        long authorization,
+        long budgetId,
+        DirectBuffer buffer,
+        int offset,
+        int limit)
+    {
+        int progress = offset;
+
+        final PgsqlMessageFW pgsqlParameter = messageRO.tryWrap(buffer, offset, limit);
+
+        if (pgsqlParameter != null)
+        {
+            client.decoder = decodePgsqlMessage;
+            progress += pgsqlParameter.length() + Byte.BYTES;
+        }
+
+        return progress;
+    }
+
+
+    private int decodePgsqlMessageType(
+        PgsqlClient client,
+        long traceId,
+        long authorization,
+        long budgetId,
+        DirectBuffer buffer,
+        int offset,
+        int limit)
+    {
+        int progressOffset = offset;
+
+        final PgsqlMessageFW pgsqlType = messageRO.tryWrap(buffer, offset, limit);
+
+        if (pgsqlType != null)
+        {
+            progressOffset = pgsqlType.limit();
+
+            short fieldCount = buffer.getShort(progressOffset, BIG_ENDIAN);
+            progressOffset += Short.BYTES;
+
```

Comment:
Does this approach work if pgsql message wraps but the remaining fields following do not?

I think we need to make sure that the progress is only advanced in the returned value if we're moving to a different decoder state, otherwise we would attempt reentry here but not aligned with the decoding logic.

Reply:
yeah so I thought about this since I don't know if it succeeds to fully decode the message I have below logic at the end of the if condition

```
if (columns.fieldCount() == fieldCount)
            {
                client.onDecodeMessageType(traceId, authorization, columns);

                client.decoder = decodePgsqlMessage;
            }
```